src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageViewDataAttributeFilter.cs

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Mvc.ViewFeatures.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages
+{
+    internal class PageViewDataAttributeFilter : IPageFilter, IPageExecutionCallbackFeature, IViewComponentExecutionCallbackFeature
+    {
+        private readonly IReadOnlyList<LifecycleProperty> _properties;
+
+        public PageViewDataAttributeFilter(IReadOnlyList<LifecycleProperty> properties)
+        {
+            _properties = properties;
+        }
+
+        public object Subject { get; set; }
+
+        public void OnPageHandlerExecuted(PageHandlerExecutedContext context)
+        {
+        }
+
+        public void OnPageHandlerExecuting(PageHandlerExecutingContext context)
+        {
+            Subject = context.HandlerInstance;
+            context.HttpContext.Features.Set<IPageExecutionCallbackFeature>(this);
+            context.HttpContext.Features.Set<IViewComponentExecutionCallbackFeature>(this);
```

Comment:
This doesn't compose by the way. You can only have one of these if it's used like this.

Reply:
They're meant to be internal contracts for now and we don't have a reason to compose them as yet. Do you foresee a reason to make them composable?